src/renderer.c

```diff
@@ -59,213 +65,206 @@ static const char* utf8_to_codepoint(const char *p, unsigned *dst) {
-  const int surface_scale = renwin_surface_scale(&window_renderer);
-  RenFont *font = font_desc_get_font_at_scale(font_desc, surface_scale);
-  while (*p) {
-    p = utf8_to_codepoint(p, &codepoint);
-    GlyphSet *set = get_glyphset(font, codepoint);
-    FR_Bitmap_Glyph_Metrics *g = &set->glyphs[codepoint & 0xff];
-    x += g->xadvance;
+int ren_font_get_width(RenFont *font, const char *text) {
+  int width = 0;
+  const char* end = text + strlen(text);
+  while (text < end) {
+    unsigned int codepoint;
+    text = utf8_to_codepoint(text, &codepoint);
+    width += font_get_glyphset(font, codepoint)->metrics[codepoint % 256].xadvance;
   }
-  /* At this point here x is in subpixel units */
-  const int x_scale_to_points = FR_Subpixel_Scale(font->renderer) * surface_scale;
-  if (subpixel_scale) {
-    *subpixel_scale = x_scale_to_points;
-    return x;
-  }
-  return (x + x_scale_to_points / 2) / x_scale_to_points;
+  return width;
 }
 
-
-int ren_get_font_height(FontDesc *font_desc) {
-  const int surface_scale = renwin_surface_scale(&window_renderer);
-  RenFont *font = font_desc_get_font_at_scale(font_desc, surface_scale);
-  return (font->height + surface_scale / 2) / surface_scale;
+float ren_font_get_size(RenFont *font) {
+  return font->size;
+}
+int ren_font_get_height(RenFont *font) {
+  return font->size + 3;
 }
 
+int ren_draw_text(RenFont *font, const char *text, float x, int y, RenColor color) {
+  SDL_Surface *surface = renwin_get_surface(&window_renderer);
+  const RenRect clip = window_renderer.clip;
+  float pen_x = x;
+  int bytes_per_pixel = surface->format->BytesPerPixel;
+  const char* end = text + strlen(text);
+  unsigned char* destination_pixels = surface->pixels;
+  int clip_end_x = clip.x + clip.width, clip_end_y = clip.y + clip.height;
+  while (text < end) {
+    unsigned int codepoint, r, g, b;
+    text = utf8_to_codepoint(text, &codepoint);
+    GlyphSet* set = font_get_glyphset(font, codepoint);
+    GlyphMetric* metric = &set->metrics[codepoint % 256];
+    int bitmap_index = font->subpixel ? (int)(fmod(pen_x, 1.0) * SUBPIXEL_BITMAPS_CACHED) : 0;
+    unsigned char* source_pixels = set->surface[bitmap_index]->pixels;
+    int start_x = pen_x + metric->bitmap_left, endX = metric->x1 - metric->x0 + pen_x;
+    int glyph_end = metric->x1, glyphStart = metric->x0;
+    if (color.a > 0 && endX >= clip.x && start_x <= clip_end_x) {
+      for (int line = metric->y0; line < metric->y1; ++line) {
+        int targetY = line + y - metric->y0 - metric->bitmap_top + font->size;
+        if (targetY < clip.y)
+          continue;
+        if (targetY >= clip_end_y)
+          break;
+        unsigned int* destination_pixel = (unsigned int*)&destination_pixels[surface->pitch * targetY + start_x * bytes_per_pixel];
+        unsigned char* source_pixel = &source_pixels[line * set->surface[bitmap_index]->pitch + metric->x0 * (font->subpixel ? 3 : 1)];
+        for (int x = glyphStart; x < glyph_end; ++x) {
```

Comment:
I think you should check if `x` is beyond the clip here too.

Reply:
This is a good point. We shouldn't draw the glyph past the clip. Will modify.

Reply:
Finally fixed this.